tls/s2n_early_data_io.c

```diff
@@ -117,3 +119,117 @@ S2N_RESULT s2n_early_data_validate_recv(struct s2n_connection *conn)
+        return S2N_RESULT_OK;
+    }
+
+    /* Attempt to make progress in the handshake even if s2n_send eventually fails.
+     * We only care about the result of this call if it would prevent us from calling s2n_send. */
+    if (s2n_negotiate(conn, blocked) < S2N_SUCCESS) {
+        if (s2n_error_get_type(s2n_errno) != S2N_ERR_T_BLOCKED) {
+            return S2N_RESULT_ERROR;
+        } else if (*blocked != S2N_BLOCKED_ON_EARLY_DATA && *blocked != S2N_BLOCKED_ON_READ) {
+            return S2N_RESULT_ERROR;
+        }
+    }
+
+    /* Attempt to send the early data.
+     * We only care about the result of this call if it fails. */
+    uint32_t early_data_to_send = 0;
+    RESULT_GUARD_POSIX(s2n_connection_get_remaining_early_data_size(conn, &early_data_to_send));
+    early_data_to_send = MIN(data_len, early_data_to_send);
+    if (early_data_to_send) {
+        ssize_t send_result = s2n_send(conn, data, early_data_to_send, blocked);
+        RESULT_GUARD_POSIX(send_result);
+        *data_sent = send_result;
+    }
+
+    /* Call s2n_negotiate again to return the current state of the handshake. */
+    if (s2n_negotiate(conn, blocked) < S2N_SUCCESS) {
+        if (s2n_error_get_type(s2n_errno) != S2N_ERR_T_BLOCKED) {
+            return S2N_RESULT_ERROR;
+        } else if (*blocked == S2N_BLOCKED_ON_EARLY_DATA) {
+            return S2N_RESULT_OK;
+        } else if (s2n_early_data_can_continue(conn)) {
+            return S2N_RESULT_ERROR;
+        } else {
+            return S2N_RESULT_OK;
+        }
+    }
+    return S2N_RESULT_OK;
+}
+
+int s2n_send_early_data(struct s2n_connection *conn, const uint8_t *data, ssize_t data_len,
+        ssize_t *data_sent, s2n_blocked_status *blocked)
+{
+    POSIX_GUARD(s2n_connection_set_early_data_expected(conn));
+    s2n_result result = s2n_send_early_data_impl(conn, data, data_len, data_sent, blocked);
+    POSIX_GUARD(s2n_connection_set_end_of_early_data(conn));
+    POSIX_GUARD_RESULT(result);
+    return S2N_SUCCESS;
+}
+
+S2N_RESULT s2n_recv_early_data_impl(struct s2n_connection *conn, uint8_t *data, ssize_t max_data_len,
+        ssize_t *data_received, s2n_blocked_status *blocked)
+{
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(blocked);
+    *blocked = S2N_NOT_BLOCKED;
+    RESULT_ENSURE_REF(data_received);
+    *data_received = 0;
+
+    RESULT_ENSURE(conn->mode == S2N_SERVER, S2N_ERR_CLIENT_MODE);
+
+    if (!s2n_early_data_can_continue(conn)) {
+        return S2N_RESULT_OK;
+    }
+
+    while(s2n_negotiate(conn, blocked) < S2N_SUCCESS) {
+        if (s2n_error_get_type(s2n_errno) != S2N_ERR_T_BLOCKED) {
+            return S2N_RESULT_ERROR;
+        } else if (max_data_len <= *data_received) {
+            return S2N_RESULT_ERROR;
+        } else if (*blocked != S2N_BLOCKED_ON_EARLY_DATA) {
+            if (s2n_early_data_can_continue(conn)) {
+                return S2N_RESULT_ERROR;
+            } else {
+                return S2N_RESULT_OK;
```

Comment:
Should we clear the blocked status and errno?

Reply:
Resetting blocked is a good call, but we don't make any promises about errno / s2n_errno on a success.